lib/chef/provider/aws_security_group.rb

```diff
@@ -1,83 +1,288 @@
+    end
+
+    #
+    # Get the list of permissions to add and remove
+    #
+    actual_rules.each do |port_range, actors|
+      if desired_rules[port_range]
+        intersection = actors & desired_rules[port_range]
+        # Anything unhandled in desired_rules will be added
+        desired_rules[port_range] -= intersection
+        # Anything unhandled in actual_rules will be removed
+        actual_rules[port_range] -= intersection
+      end
+    end
+
+    #
+    # Add any new rules
+    #
+    desired_rules.each do |port_range, actors|
+      unless actors.empty?
+        authorize.call(port_range[:port_range], port_range[:protocol], actors)
+      end
+    end
+
+    #
+    # Remove any rules no longer in effect
+    #
+    actual_rules.each do |port_range, actors|
+      unless actors.empty?
+        revoke.call(port_range[:port_range], port_range[:protocol], actors)
+      end
+    end
+  end
+
+  def add_rule(rules, port_ranges, actors)
+    unless actors.empty?
+      port_ranges.each do |port_range|
+        rules[port_range] ||= Set.new
+        rules[port_range] += actors
       end
     end
   end
 
-  # TODO need support for load balancers!
-  def get_sources(sources)
-    sources.map do |s|
-      if s.is_a?(String)
-        begin
-          IPAddr.new(s)
-          s
-        rescue
-          { group_id: Chef::Resource::AwsSecurityGroup.get_aws_object_id(s, resource: new_resource) }
+  def get_port_ranges(port_spec)
+    case port_spec
+    when Integer
+      [ { port_range: port_spec..port_spec, protocol: :tcp } ]
+    when Range
+      [ { port_range: port_spec, protocol: :tcp } ]
+    when Array
+      port_spec.map { |p| get_port_ranges(p) }.flatten
+    when Hash
+      if port_spec[:protocol]
+        [ { port_range: port_spec[:port_range] || port_spec[:port], protocol: port_spec[:protocol].to_s.to_sym } ]
+      else
+        get_port_ranges(port_spec[:port_range] || port_spec[:port])
+      end
+      # The to_s.to_sym dance is because if you specify a protocol number, AWS symbolifies it,
+      # but 26.to_sym doesn't work (so we have to to_s it first).
+    when nil
+      [ { port_range: nil, protocol: :any } ]
+    end
+  end
+
+  #
+  # Turns an actor_spec into a uniform array, containing CIDRs, AWS::EC2::LoadBalancers and AWS::EC2::SecurityGroups.
+  #
+  def get_actors(vpc, actor_spec)
+    result = case actor_spec
+
+    # An array is always considered a list of actors.  Each one may follow any supported format.
+    when Array
+      actor_spec.map { |a| get_actors(vpc, a) }
+
+    # Hashes come in several forms:
+    when Hash
+      # The default AWS Ruby SDK form with :user_id, :group_id and :group_name forms
+      if actor_spec.keys.all? { |key| [ :user_id, :group_id, :group_name ].include?(key) }
+        if actor_spec.has_key?(:group_name)
+          actor_spec[:group_id] ||= vpc.security_groups.filter('group-name', actor_spec[:group_name]).first.id
```

Comment:
I was lying here - vpc is passed in and it is provided from the AWS object, so it cannot be nil